### ISSUE TITLE
Could not unroll loop. Loop bound could not be deduced at compile time

### ISSUE THREAD

**Issue by jirzynek** [microsoft/DirectXShaderCompiler#2269]
DXC fails to compile the following code:
```
// dxc.exe test2.hlsl -nologo -O3 -Zi -Qstrip_debug -Tps_5_1 -Emain_ps -Fo"test.bin"

#define MM 2
#define KK 3
#define NN 1    // change from 1 to 2 and no errors

float c1()
{
    float a = 1.f;
    
    [unroll]
    for( uint i = 0 ; i < MM && a > 1e-4; ++i )
    //for( uint i = 0 ; i < MM;  ++i )          // uncomment this -> no errors
    {
        float b = 0.01;     // comment this out -> no errors
        a *= 1.0f - b;
    }
    return 0;
}

float c3()
{
    [unroll] for (int i = 0; i < NN; ++i)         // comment this out -> no errors
    {
        [unroll] for (int j = 0; j < (KK - 1); ++j)     // comment this out -> no errors
        {
        }
    }
    
	return 0;
}

float main_ps() : SV_Target
{
    return c1() * c3();
}

```
and it produces the following error:

```
Could not unroll loop. Loop bound could not be deduced at compile time. Use [unroll(n)] to give an explicit count. Use '-HV 2016' to treat this as warning.

```

FXC compiles the sample without errors.

**Comment by MrTrillian** [comment]
Thanks for reporting! This is a bit of a bizarre repro, it's not clear why the product of those functions would behave differently from them taking into isolation. However in general we cannot garantee that we can unroll loops when the bound requires the compiler doing arbitrary math. We just do a best effort for backcompat with FXC because it would accept some surprising patterns, but it is impossible for us to match behavior in all cases.

@adam-yang is our local expert. Adam, do you think this specific case should be tracked as a bug?

**Comment by adam-yang** [comment]
Fixed now with #2291
